src/irmin-pack/unix/file_manager.ml

```diff
@@ -558,12 +571,44 @@ struct
       Control.set_payload t.control pl
     in
 
-    (* Step 3. Use the new suffix in the rw instance *)
+    (* Step 4. Use the new suffix in the rw instance. *)
     let old_suffix = t.suffix in
-    t.suffix <- suffix;
+    t.suffix <- new_suffix;
 
-    (* Step 4. Close, unlink and return *)
+    (* Step 5. Close, unlink and return *)
     let* () = Suffix.close old_suffix in
     let* () = if unlink then Io.unlink (Suffix.path old_suffix) else Ok () in
     Ok ()
+
+  let write_gc_output ~root ~generation output =
+    let open Result_syntax in
+    let path = Irmin_pack.Layout.V3.gc_result ~root ~generation in
+    let* io = Io.create ~path ~overwrite:true in
+    let out = Errs.to_json_string output in
+    let* () = Io.write_string io ~off:Int63.zero out in
+    Io.close io
+
+  let read_gc_output ~unlink ~root ~generation =
+    let open Result_syntax in
+    let read_file () =
+      let path = Irmin_pack.Layout.V3.gc_result ~root ~generation in
+      let* io = Io.open_ ~path ~readonly:true in
+      let* len = Io.read_size io in
+      let len = Int63.to_int len in
+      let* string = Io.read_to_string io ~off:Int63.zero ~len in
+      let* () = Io.close io in
+      let* () = if unlink then Io.unlink path else Ok () in
+      Ok string
+    in
+    match read_file () with
+    | Error err -> Error (`Corrupted_gc_result_file (Fmt.str "%a" Errs.pp err))
+    | Ok x -> (
+        match Errs.of_json_string x with
+        | Ok (generation', last_offset) when generation' = generation ->
+            Ok last_offset
+        | Ok (generation', _) ->
+            Error
+              (`Gc_process_error
+                (Fmt.str "Expected generation %d, got %d" generation generation'))
```

Comment:
Now that `generation` is saved in `repo`. It is no longer necessary to return it from `Gc.run`. Removing it from there will avoid this error case.